controller/src/main/java/io/pravega/controller/server/SegmentStoreConnectionManager.java

```diff
@@ -0,0 +1,653 @@
+        }
+    }
+    
+    static class ConnectionObject {
+        private final ClientConnection connection;
+        private final ReusableReplyProcessor reusableReplyProcessor;
+        private final AtomicReference<ConnectionState> state;
+
+        ConnectionObject(ClientConnection connection, ReusableReplyProcessor processor) {
+            this.connection = connection;
+            this.reusableReplyProcessor = processor;
+            state = new AtomicReference<>(ConnectionState.CONNECTED);
+        }
+
+        ConnectionState getState() {
+            return state.get();
+        }
+
+        void failConnection() {
+            state.set(ConnectionState.DISCONNECTED);    
+        }
+        
+        <T> void sendAsync(WireCommand request, CompletableFuture<T> resultFuture) {
+            connection.sendAsync(request, cfe -> {
+                if (cfe != null) {
+                    Throwable cause = Exceptions.unwrap(cfe);
+                    if (cause instanceof ConnectionFailedException) {
+                        resultFuture.completeExceptionally(new WireCommandFailedException(cause, request.getType(),
+                                WireCommandFailedException.Reason.ConnectionFailed));
+                        state.set(ConnectionState.DISCONNECTED);
+                    } else {
+                        log.debug("connection.sendAsync failed with {}", cause.getClass());
+                        resultFuture.completeExceptionally(cause);
+                    }
+                }
+            });
+        }
+
+        private enum ConnectionState {
+            CONNECTED,
+            DISCONNECTED
+        }
+    }
+
+    @VisibleForTesting
+    static interface ConnectionListener {
+        enum ConnectionEvent {
+            NewConnection,
+            ConnectionClosed
+        }
+        
+        void notify(ConnectionEvent event);
+    }
+    
+    @Data
+    private static class WaitingRequest {
+        private final CompletableFuture<ConnectionObject> future;
+        private final ReplyProcessor replyProcessor;
+    }
+    
+    /**
+     *  A reusable reply processor class which can be initialized and uninitialized with new ReplyProcessor. 
+     *  This same replyProcessor can be reused with the same connection for handling different replies from servers for
+     *  different calls.
+     */
+    private static class ReusableReplyProcessor implements ReplyProcessor {
+        private final AtomicReference<ReplyProcessor> replyProcessor = new AtomicReference<>();
+
+        // initialize the reusable reply processor class with a new reply processor
+        void initialize(ReplyProcessor replyProcessor) {
+            this.replyProcessor.set(replyProcessor);
+        }
+
+        // unset reply processor
+        void uninitialize() {
+            replyProcessor.set(null);
+        }
+
+        @Override
+        public void hello(WireCommands.Hello hello) {
+            ReplyProcessor rp = replyProcessor.get();
```

Comment:
Can you just write a method that does this pattern instead of rewriting the same code over and over again?

```
    private <T> void execute(BiConsumer<ReplyProcessor,T> toInvoke, T arg){
        ReplyProcessor rp = replyProcessor.get();
        if (rp != null) {
            toInvoke.accept(rp, arg);
        }
    }
```

then you call it like this:
```
        execute(ReplyProcessor::hello, hello);
```